dashboard/app/assets/javascripts/callouts.js

```diff
@@ -1,12 +1,53 @@
-dashboard.createCallouts = function(callouts) {
-  if (!callouts) {
-    return;
+(function () {
+
+  /**
+   * Snap all callouts to their target positions.  Keeps them in
+   * position when blockspace is scrolled.
+   */
+  function snapCalloutsToTargets() {
+    var triggerEvent = null;
+    var animate = false;
+    $('.cdo-qtips').qtip('reposition', triggerEvent, animate);
   }
 
-  // Hide callouts when the function editor is closed (otherwise they jump to the top left corner)
-  $(window).on('function_editor_closed', function() {
-    $('.cdo-qtips').qtip('hide');
-  });
+  /**
+   * For callouts with targets in the codeWorkspace (blockly, flyout elements,
+   * function editor elements, etc) hides callouts with targets that are
+   * scrolled out of view, and shows them again when they are scrolled back in
+   * to view.
+   * @function
+   */
+  var showOrHideCalloutsByTargetVisibility = (function () {
+    // Close around this object, which we use to remember which callouts
+    // were hidden by scrolling and should be shown again when they scroll
+    // back in.
+    /** @type {Object.<string, boolean>} */
+    var calloutsHiddenByScrolling = {};
+    return function () {
+      var codeWorkspace = $('#codeWorkspace');
```

Comment:
The code will below will get a NPE if $('#codeWorkspace') doesn't exist.

I believe this change makes this code more tightly bind to a particular page structure than it was previously; prior to this point we just need to have 0 or more element with class cdo-qtips.

Please consider whether we want this code to require a particular page structure or whether we might want to reuse it with other pages.  If we do want to require that structure, please add a comment but also have the code gracefully handle the case where the div is missing.  If you want to keep the code more general, perhaps you could move the block_space_metrics_set event handler to a page specific javascript file. (Sorry, I don't know the structure of the code well enough to know if that is a feasible suggestion.)